streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -38,20 +48,36 @@
     private final String otherWindowName;
     private final long joinBeforeMs;
     private final long joinAfterMs;
+    private final long joinGraceMs;
 
     private final ValueJoinerWithKey<? super K, ? super V1, ? super V2, ? extends R> joiner;
     private final boolean outer;
+    private final Optional<String> outerJoinWindowName;
+    private final boolean thisJoin;
+
+    // Observed time is AtomicLong because this time is shared between the left and side processor nodes. However,
+    // this time is not updated in parallel, so we can call get() several times without worry about getting different
+    // times.
+    private final AtomicLong maxObservedStreamTime;
```

Comment:
Two comments:
- even if shared between two processors, both would be executed in the same thread and thus it seems not be necessary to use `AtomicLong`
- do we need to maintain it manually? could we use `contect.streamTime()` instead?

Reply:
I think `context.currentStreamTimeMs()` should work. I wasn't awrare I could get the new stream time from it. I don't see any problems as I only need the stream time to expire the records.

Reply:
We had similar debate for using per-task stream time v.s. per-processor stream time when it comes down to expiration. I'm concerned if we go with task stream time here we would have an inconsistent behavior compared with other scenarios?

Reply:
Agreed, but I am wondering if we should/want to switch to the new `context.streamTime()` everywhere? -- It's totally open question, but might be worth to explore.

Reply:
I personally was on the side of always using task stream time everywhere but more people feel that we should use processor stream time :P Anyways, all I'm trying to say is that we need to make an educated decision here, and if we concluded that either 1) we rely on task time here, but still use processor time on other expiration logic, or 2) we rely on processor time on all logic, or 3) we rely on task time on all logic, we have a good rationale for whichever we choose.